sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala

```diff
@@ -28,6 +33,57 @@ import org.apache.spark.util.Utils
 /**
  * Helper classes for reading/writing state schema.
  */
+sealed trait ColumnFamilySchema extends Serializable {
+  def jsonValue: JValue
+
+  def json: String
+
+  def columnFamilyName: String
+}
+
+case class ColumnFamilySchemaV1(
+    columnFamilyName: String,
+    keySchema: StructType,
+    valueSchema: StructType,
+    keyStateEncoderSpec: KeyStateEncoderSpec,
+    userKeyEncoder: Option[StructType] = None) extends ColumnFamilySchema {
```

Comment:
I feel like userKeyEncoder is too coupled with transformWithState, but I couldn't imagine we would have more complicated case of stateful operator implementation, so sounds like OK to me.